src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { take } from 'rxjs/operators';
+import { compileFile } from 'pug';
+import { join } from 'path';
+
+import { i18n } from '@kbn/i18n';
+
+// TODO: Replace once https://github.com/elastic/kibana/pull/51438 lands:
+import { Capabilities } from '../../../core/public';
+import { Env } from '../config';
+import { InternalHttpServiceSetup, KibanaRequest, KibanaResponseFactory } from '../http';
+import { IUiSettingsClient } from '../ui_settings';
+import { PluginsServiceSetup, DiscoveredPlugin } from '../plugins';
+import { createCspRuleString } from '../csp';
+import { mergeVariables } from './merge_variables';
+import { IApplicationContextProvider, PluginVariables, UserVariables } from './types';
+
+interface Params {
+  request: KibanaRequest;
+  response: KibanaResponseFactory;
+  env: Env;
+  http: InternalHttpServiceSetup;
+  uiSettings: IUiSettingsClient;
+  plugins: PluginsServiceSetup;
+}
+
+interface ConfiguredPlugin {
+  id: string;
+  plugin: DiscoveredPlugin;
+  config: unknown;
+}
+
+const template = compileFile(join(__dirname, 'views/ui_app.pug'));
+// TODO: Remove once https://github.com/elastic/kibana/pull/51438 lands:
+const defaultCapabilities: Capabilities = {
+  navLinks: {},
+  management: {},
+  catalogue: {},
+};
+
+export class ApplicationContextProvider implements IApplicationContextProvider {
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: IUiSettingsClient;
+  private readonly plugins!: PluginsServiceSetup;
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
```

Comment:
I think it's preferable to keep type safety here and not use `Object.assign` combined with `!` on these dependencies. Instead, can we use a similar pattern used in other services:

```ts
constructor(private readonly params: Params) {}
```

And then use `this.params.*` everywhere else.